drivers/crypto/crypto_si32.c

```diff
@@ -0,0 +1,1170 @@
+{
+	struct crypto_session *session;
+	int ret;
+
+	if (!ctx) {
+		LOG_WRN("Missing context");
+		return -EINVAL;
+	}
+
+	session = (struct crypto_session *)ctx->drv_sessn_state;
+
+	if (!pkt) {
+		LOG_WRN("Missing packet");
+		return -EINVAL;
+	}
+
+	if (pkt->in_len % 16) {
+		LOG_ERR("Can't work on partial blocks");
+		return -EINVAL;
+	}
+
+	if (pkt->in_len > 16) {
+		LOG_ERR("Refusing to work on multiple ECB blocks");
+		return -EINVAL;
+	}
+
+	if (pkt->in_len == 0) {
+		LOG_DBG("Zero-sized packet");
+		return 0;
+	}
+
+	if ((ctx->flags & CAP_INPLACE_OPS) && (pkt->out_buf != NULL)) {
+		LOG_ERR("In-place must not have an out_buf");
+		return -EINVAL;
+	}
+
+	/* As per 12.6.1./12.6.2. Configuring the DMA for ECB Encryption/Decryption */
+
+	/* DMA Input Channel */
+	ret = crypto_si32_dma_setup_tx(pkt, 0);
+	if (ret) {
+		return ret;
+	}
+
+	/* DMA Output Channel */
+	ret = crypto_si32_dma_setup_rx(pkt, 0, 0);
+	if (ret) {
+		return ret;
+	}
+
+	/* AES Module */
+
+	/* 1. The XFRSIZE register should be set to N-1, where N is the number of 4-word blocks. */
+	SI32_AES_A_write_xfrsize(SI32_AES_0, pkt->in_len / AES_BLOCK_SIZE - 1);
+
+	switch (op) {
+	case CRYPTO_CIPHER_OP_ENCRYPT:
+		/* 2. The HWKEYx registers should be written with the desired key in little endian
+		 * format.
+		 */
+		ret = crypto_si32_aes_set_key(ctx->key.bit_stream, ctx->keylen);
+		if (ret) {
+			return ret;
+		}
+		break;
+	case CRYPTO_CIPHER_OP_DECRYPT:
+		/* 2. The HWKEYx registers should be written with decryption key value
+		 * (automatically generated in the HWKEYx registers after the encryption process).
+		 */
+		ret = crypto_si32_aes_set_key(session->decryption_key, ctx->keylen);
+		if (ret) {
+			return ret;
+		}
+		break;
+	}
+
+	/* 3. The CONTROL register should be set as follows: */
+	{
+		__ASSERT(SI32_AES_0->CONTROL.ERRIEN == 1, "a. ERRIEN set to 1.");
+
+		/* KEYSIZE set to the appropriate number of bits for the key. */
+		ret = crypto_si32_aes_set_key_size(ctx);
+		if (ret) {
+			return ret;
+		}
+
+		switch (op) {
+		/* c. EDMD set to 1 for encryption. */
+		case CRYPTO_CIPHER_OP_ENCRYPT:
+			SI32_AES_A_select_encryption_mode(SI32_AES_0);
+			break;
+		/* c. EDMD set to 1 for DEcryption. (documentation is wrong here) */
+		case CRYPTO_CIPHER_OP_DECRYPT:
+			SI32_AES_A_select_decryption_mode(SI32_AES_0);
+			break;
+		}
+
+		/* d. KEYCPEN set to 1 to enable key capture at the end of the transaction. */
+		SI32_AES_A_enable_key_capture(SI32_AES_0);
+
+		/* e. The HCBCEN, HCTREN, XOREN, BEN, SWMDEN bits should all be cleared to 0. */
+		SI32_AES_A_exit_cipher_block_chaining_mode(SI32_AES_0); /* Clear HCBCEN */
+		SI32_AES_A_exit_counter_mode(SI32_AES_0);               /* Clear HCTREN */
+		SI32_AES_A_select_xor_path_none(SI32_AES_0);            /* Clear XOREN */
+		SI32_AES_A_exit_bypass_hardware_mode(SI32_AES_0);       /* Clear BEN */
+		SI32_AES_A_select_dma_mode(SI32_AES_0);                 /* Clear SWMDEN*/
+	}
+
+	k_sem_reset(&crypto_si32_work_done);
```

Comment:
I think this code here itself is fine, because the DMA setup procedure should cancel all potentially running DMA operations.

I'll have to revisit the DMA code however.

Reply:
I'm not yet 100% it is a good thing, but this is my current attempt at improving the DMA code: #79953

Reply:
@ceolin If you happen to have some time/nerves, I'd appreciate your feedback on #79953. Thanks!

Reply:
Just pushed a (I think) improved version that makes it clear that DMA is not running when resetting the semaphore.